### PR TITLE
[NO SQUASH] Mod storage PostgreSQL backend

### PR DESCRIPTION
**Pull request by TurkeyMcMac**
Closes https://github.com/minetest/minetest/issues/12873. It seems server owners want a PostgreSQL backend. I have implemented `getModKeys` from https://github.com/minetest/minetest/pull/12841 for compatibility with that PR.

## To do

This PR is Ready for Review.

## How to test

Ideally you should test with Postgres versions before and after 9.5.

1. Run the DevTest unit tests.
2. Run the C++ unit tests with `MINETEST_POSTGRESQL_CONNECT_STRING` set appropriately.

### CODE REVIEW

**Comment by sfan5** [comment]
We *can* reasonably unittest this, I propose the following:
1. in the code attempt to retrieve e.g. `POSTGRES_CONNECT_STRING` from env, skip test otherwise
2. during CI start a docker container that has pg
3. export the connect string in CI

**Comment by TurkeyMcMac** [comment]
IMO the PostgreSQL backends should be disabled for systems with 16-bit integers. `libpq` uses the `int` type in many places. If it is only 16 bits, `MapDatabasePostgreSQL::listAllLoadableBlocks` can only return up to 32767 blocks, not nearly enough. There are other truncation problems elsewhere.

Thoughts?

**Comment by nerzhul** [comment]
who play minetest on a 16byte computer ??

**Comment by TurkeyMcMac** [comment]
The C standard allows `int` to be 16 bits. Perhaps the global check can be added in a separate PR.

**Comment by sfan5** [comment]
Not sure why this comes up now, we have worse issues if int is fewer than 32-bits.

File: src/database/database-postgresql.cpp [hunk 1]
```diff
@@ -812,5 +813,210 @@ void AuthDatabasePostgreSQL::writePrivileges(const AuthEntry &authEntry)
...
+			"INSERT INTO mod_storage (modname, key, value) VALUES ($1, $2::bytea, $3::bytea) "
+				"ON CONFLICT ON CONSTRAINT mod_storage_pkey DO "
+					"UPDATE SET value = $3::bytea");
+	}
+	prepareStatement("remove",
+		"DELETE FROM mod_storage WHERE modname = $1 AND key = $2::bytea");
+	prepareStatement("remove_all",
+		"DELETE FROM mod_storage WHERE modname = $1");
+	prepareStatement("list",
+		"SELECT modname FROM mod_storage GROUP BY modname");
```
Comment (sfan5) [inline src/database/database-postgresql.cpp:866]:
```suggestion
		"SELECT DISTINCT modname FROM mod_storage");
```

File: src/database/database-postgresql.cpp [hunk 2]
```diff
@@ -812,5 +813,210 @@ void AuthDatabasePostgreSQL::writePrivileges(const AuthEntry &authEntry)
...
+	const void *args[] = { modname.c_str() };
+	const int argLen[] = { -1 };
+	const int argFmt[] = { 0 };
+	PGresult *results = execPrepared("get_all", ARRLEN(args),
+			args, argLen, argFmt, false);
+
+	int numrows = PQntuples(results);
+
+	for (int row = 0; row < numrows; ++row) {
+		std::string key(PQgetvalue(results, row, 0), PQgetlength(results, row, 0));
```
Comment (sfan5) [inline src/database/database-postgresql.cpp:882]: would be wise to add a helper for this

File: src/database/database-postgresql.cpp [hunk 3]
```diff
@@ -812,5 +813,210 @@ void AuthDatabasePostgreSQL::writePrivileges(const AuthEntry &authEntry)
...
+bool ModMetadataDatabasePostgreSQL::setModEntry(const std::string &modname,
+	const std::string &key, const std::string &value)
+{
+	verifyDatabase();
+
+	const void *args[] = { modname.c_str(), key.c_str(), value.c_str() };
+	const int argLen[] = {
+		-1,
+		(int)MYMIN(key.size(), (size_t)INT_MAX),
+		(int)MYMIN(value.size(), (size_t)INT_MAX),
```
Comment (sfan5) [inline src/database/database-postgresql.cpp:961]: suggestion:
```cpp
template<typename T, typename = std::enable_if_t<std::is_unsigned<T>::value>>
int clamp_int(T n) {
    return static_cast<int>(MYMIN(n, INT_MAX));
}
```

File: src/unittest/test_modmetadatadatabase.cpp [hunk 4]
```diff
@@ -104,6 +114,49 @@ class SQLite3Provider : public ModMetadataDatabaseProvider
...
+			PQconnectdb(connect_string.c_str()), PQfinish);
+	if (PQstatus(conn.get()) == CONNECTION_OK) {
+		std::unique_ptr<PGresult, decltype(&PQclear)> res(
+				PQexec(conn.get(), "DROP TABLE IF EXISTS mod_storage;"), PQclear);
+		if (PQresultStatus(res.get()) == PGRES_COMMAND_OK)
+			return;
+	}
+	throw DatabaseException(
+			std::string("PostgreSQL database error: ") + PQerrorMessage(conn.get()));
+}
```
Comment (sfan5) [inline src/unittest/test_modmetadatadatabase.cpp:131]: for simplicity reasons don't we want to make this a `listMods` + `removeModEntries`?

File: src/database/database-postgresql.cpp (hunk 3 above)
Comment (TurkeyMcMac) [reply src/database/database-postgresql.cpp]: I have simplified the code a bit so that I don't think a helper function is necessary.

**Review by sfan5 — APPROVED** [review]

**Review by nerzhul — APPROVED** [review]

**Pull request merged**